packages/components/src/side_scroller/index.jsx

```diff
@@ -0,0 +1,136 @@
+  const style = { height };
+
+  const getScrollableNode = () => containerRef.current;
+
+  const stopScrollAnimation = () => {
+    if (animationIdRef.current) global.cancelAnimationFrame(animationIdRef.current);
+  };
+
+  const updateScroll = () => {
+    const { scrollLeft } = containerRef.current;
+    const maxScrollPosition = contentWidthRef.current - containerWidthRef.current;
+
+    const nextCanScrollLeft = scrollLeft > 0;
+    const nextCanScrollRight = scrollLeft < maxScrollPosition;
+
+    setCanScrollLeft(nextCanScrollLeft);
+    setCanScrollRight(nextCanScrollRight);
+  };
+
+  const updateSizes = () => {
+    const { width: containerWidth } = size(containerRef.current);
+    const { height: nextHeight } = size(contentRef.current);
+
+    containerWidthRef.current = containerWidth;
+    contentWidthRef.current = getScrollableNode().scrollWidth;
+
+    updateScroll();
+    setHeight(nextHeight);
+  };
+
+  const handleDragStart = (event) => {
+    // Prevents DOM nodes like images from being dragged
+    event.preventDefault();
+    startXRef.current = eventCoordinates(event, 'pageX').pageX;
+    startPositionRef.current = getScrollableNode().scrollLeft;
+    stopScrollAnimation();
+  };
+
+  const handleDrag = (event) => {
+    const diff = startXRef.current - eventCoordinates(event, 'pageX').pageX;
+    const shift = Math.abs(diff);
+    const newPosition = startPositionRef.current + diff;
+
+    if (shift >= DISABLE_SCROLL_DISTANCE) {
+      preventClickRef.current = true;
+
+      // Prevents vertical scroll on touch devices
+      event.preventDefault();
+    }
+
+    getScrollableNode().scrollLeft = newPosition;
+  };
+
+  const handleClickCapture = (event) => {
+    if (preventClickRef.current) {
+      stopEvent(event);
+      preventClickRef.current = false;
+    }
+  };
+
+  const handleAnimateScroll = (callback) => {
+    animationIdRef.current = global.requestAnimationFrame(callback);
+  };
+
+  useEffect(() => {
+    stopListeningToResizeRef.current = eventproxy('resize', updateSizes);
+    updateSizes();
+
+    return () => {
+      stopListeningToResizeRef.current();
+      stopScrollAnimation();
+    };
+  }, []);
+
+  useEffect(() => {
+    updateSizes();
+  }, [children]);
+
+  return (
+    <article {...cleanProps} className={className}>
+      <div
+        className={styles.container} ref={containerRef} style={style}
+        onScroll={updateScroll} onLoad={updateSizes}
+      >
+        <Draggable
+          contentRef={contentRef}
+          onDragStart={handleDragStart}
+          onDrag={handleDrag}
+          onClickCapture={handleClickCapture}
+        >
+          {children}
+        </Draggable>
+      </div>
+      <Controls
+        canScrollLeft={canScrollLeft}
+        canScrollRight={canScrollRight}
+        scrollableNode={getScrollableNode()}
+        animationIdRef={animationIdRef}
+        containerWidthRef={containerWidthRef}
+        contentWidthRef={contentWidthRef}
```

Comment:
We cannot pass around so many refs. That somehow defies the whole component based dataflow logic. Can we extract the scrolling into this component? @ozerovs